### ISSUE TITLE
Surprising datetime autoscaling after passing empty data

### ISSUE THREAD

**Issue by mwaskom** [matplotlib/matplotlib#17586]
### Bug report

**Bug summary**

The automatic limits for datetime axes depend in surprising ways on the presence of artists without any data.

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt
```

Case 1:

If a datetime plot follows an empty plot, the autoscaling on the date axis are not what one would expect:

```python
f, ax = plt.subplots()
x = np.arange("2010-01-01", "2011-01-01", dtype="datetime64[D]")
ax.scatter([], [])
ax.scatter(x, np.arange(len(x)))
```
![image](https://user-images.githubusercontent.com/315810/83967872-38832880-a893-11ea-84fe-b8a0661cb700.png)

While I can't account for the specific choice of limits, I gather that the first plot sets the units on the axes in a way that interferes with the subsequent plot. And indeed, we can fix the problem by setting a dtype on our null data.

Case 2:

```python
f, ax = plt.subplots()
x = np.arange("2010-01-01", "2011-01-01", dtype="datetime64[D]")
null = np.array([], x.dtype)
ax.scatter(null, [])
ax.scatter(x, np.arange(len(x)))
```
![image](https://user-images.githubusercontent.com/315810/83967877-4042cd00-a893-11ea-805e-4b7fa27b0757.png)

But this does not work for all functions. Consider the case of `fill_between`.

Case 3:

```python
f, ax = plt.subplots()
x = np.arange("2010-01-01", "2011-01-01", dtype="datetime64[D]")
null = np.array([], x.dtype)
ax.fill_between(null, [])
ax.fill_between(x, np.arange(len(x)))
```
![image](https://user-images.githubusercontent.com/315810/83967888-4769db00-a893-11ea-9a8e-04e1de01dd4a.png)

In summary, I think Case 1 is not a bug (it would be nice to defer unit setting for empty data, but I can easily imagine that this is hard/impossible). But I can't account for Case 3.

**Matplotlib version**

  * Operating system: macos
  * Matplotlib version: 3.2.1
  * Matplotlib backend (`print(matplotlib.get_backend())`): pylab inline
  * Python version: 3.8
